crates/runtime/src/http/v1.rs

```diff
@@ -843,3 +883,91 @@ pub(crate) mod inference {
         }
     }
 }
+
+
+// #[cfg(feature = "nsql")]
+pub(crate) mod nsql {
+    use arrow::datatypes::Schema;
+    use axum::{
+        debug_handler, http::StatusCode, response::{IntoResponse, Response}, Extension, Json
+    };
+    use futures::{future::join_all, TryFutureExt};
+    use itertools::Itertools;
+    use llms::nql::Nql;
+    use serde::{Deserialize, Serialize};
+    use std::{ops::DerefMut, pin::Pin, sync::Arc};
+    use tokio::sync::RwLock;
+    
+    use crate::{datafusion::DataFusion, http::v1::dataframe_to_response};
+
+    #[derive(Debug, Serialize, Deserialize)]
+    #[serde(rename_all = "lowercase")]
+    pub struct NsqlRequest {
+        pub query: String,
+
+        // If present, only include these tables in the NSQL query
+        // pub included_tables: Option<Vec<String>>,
+    }
+
+    pub fn schema_to_create_table_stmt(table_name: &str, schema: Schema) -> String {
+        return "".to_string();
+    }
+
+    // #[debug_handler]
+    pub(crate) async fn post(
+        Extension(df): Extension<Arc<RwLock<DataFusion>>>,
+        Extension(nsql_model):  Extension<Arc<RwLock<Pin<Box<dyn Nql>>>>>,
+        Json(payload): Json<NsqlRequest>,
+    ) -> Response {
+
+        let readable_df = df.read().await;
+
+        let tables = match readable_df.get_public_table_names().await {
+            Ok(t) => t,
+            Err(e) => {
+                tracing::trace!("Error getting tables: {e}");
+                return (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()).into_response();
+            }
+        };
+
+        match join_all(tables.iter().map(|t| {
+            readable_df.get_arrow_schema(t).map_ok(|s| {
+                schema_to_create_table_stmt(t, s)
+            }
+        )})).await.into_iter().collect::<Result<Vec<String>, _>>() {
+            Ok(create_tbl_stmts) => {
+                let nsql_query = format!(
+                    "{query} -- Using valid ANSI SQL, answer the following questions for the tables provided above. -- {user_query}",
+                    user_query=payload.query,
+                    query=create_tbl_stmts.join(" ")
+                );
+                tracing::error!("Running query: {nsql_query}");
+                
+                let mut model = nsql_model.write().await; 
```

Comment:
Key issue/bug is here. model is a `Extension<Arc<RwLock<Pin<Box<dyn Nql>>>>>`, but run needs a `mut self`. See:
```rust
#[async_trait]
pub trait Nql: Sync + Send {
    async fn run(&mut self, prompt: String) -> Result<Option<String>>;
}
```